### PR TITLE
[FIX] sale_timesheet: take credit notes into account when invoicing

### PR DESCRIPTION
**Pull request by pivi-odoo**
When invoicing a timesheet service (with selected dates that include the service's logged time), that was partially refunded/credited, more hours are invoiced than delivered.

Steps to reproduce:
- create a service that is invoiced based on delivered timesheet, based on task or project, unit hours
- create a sale order with a line of that service with a certain quantity, validate the order and generate the invoice
- create a credit note that refunds part of that invoiced quantity
- log more time in the timesheet
- create a new invoice (select dates that include the newly logged work)

Current result: more hours were invoiced than delivered Expected result: invoiced quantity should be equal to the delivered quantity

Fix:
When we are calculating how much to invoice for each line, if it is a timesheeted one, and it was previously refunded, substract the currently invoiced quantity from the line.

Affected versions:
- 15.0
- saas-15.2
- saas-15.3

⚠️ ~~as of writing this PR, the bug couldn't be reproduced on master (now version 16.0) due to another issue (the wizard doesn't show date fields when having sale order lines based on timesheeted tasks). When that issue gets resolved, forward port this PR to version 16.0 and master.~~ I've written a fix `https://github.com/odoo/odoo/pull/101020`, the issue has been confirmed present on 16.0 and master.

opw-2974259

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pivi-odoo** [comment]
@fw-bot up to saas-15.3

**Comment by pivi-odoo** [comment]
@fw-bot up to master

File: addons/sale_timesheet/models/sale_order.py [hunk 1]
```diff
@@ -280,5 +280,12 @@ def _recompute_qty_to_invoice(self, start_date, end_date):
             domain = expression.AND([domain, [('date', '<=', end_date)]])
         mapping = lines_by_timesheet.sudo()._get_delivered_quantity_by_analytic(domain)
 
+        # we need to remove the already invoiced quantity for lines that were credited
+        if refund_account_moves:
+            for sol_id in mapping:
+                line = refund_account_moves.line_ids.sale_line_ids.filtered(lambda sol: sol.id == sol_id)
+                if line:
+                    mapping[sol_id] -= line.qty_invoiced
```
Comment (xavierbol) [inline addons/sale_timesheet/models/sale_order.py:288]: And if we already invoiced something, we will remove more than expected, won't we? 🤔

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 2]
```diff
@@ -524,6 +524,171 @@ def test_timesheet_invoice(self):
...
+            'order_id': sale_order.id,
+        })
+        so_line_deliver_task_project = self.env['sale.order.line'].create({
+            'name': self.product_delivery_timesheet3.name,
+            'product_id': self.product_delivery_timesheet3.id,
+            'product_uom_qty': 50,
+            'product_uom': self.product_delivery_timesheet3.uom_id.id,
+            'price_unit': self.product_delivery_timesheet3.list_price,
+            'order_id': sale_order.id,
+        })
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:562]: You can create in batch. 🙂 
```suggestion
        so_line_deliver_global_project, so_line_deliver_task_project = self.env['sale.order.line'].create([
            {
                'name': self.product_delivery_timesheet2.name,
                'product_id': self.product_delivery_timesheet2.id,
                'product_uom_qty': 50,
                'product_uom': self.product_delivery_timesheet2.uom_id.id,
                'price_unit': self.product_delivery_timesheet2.list_price,
                'order_id': sale_order.id,
            },
            {
                'name': self.product_delivery_timesheet3.name,
                'product_id': self.product_delivery_timesheet3.id,
                'product_uom_qty': 50,
                'product_uom': self.product_delivery_timesheet3.uom_id.id,
                'price_unit': self.product_delivery_timesheet3.list_price,
                'order_id': sale_order.id,
            },
        ])
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 3]
```diff
@@ -524,6 +524,171 @@ def test_timesheet_invoice(self):
...
+            'date': today
+        })
+        self.env['account.analytic.line'].create({
+            'name': 'Test Line 2',
+            'project_id': task_serv2.project_id.id,
+            'task_id': task_serv2.id,
+            'unit_amount': 5,
+            'employee_id': self.employee_manager.id,
+            'date': today
+        })
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:631]: Same here, you can do the create in batch. 🙂

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 4]
```diff
@@ -524,6 +524,171 @@ def test_timesheet_invoice(self):
...
+        self.env['account.analytic.line'].create({
+            'name': 'Test Line 2',
+            'project_id': task_serv2.project_id.id,
+            'task_id': task_serv2.id,
+            'unit_amount': 15,
+            'employee_id': self.employee_manager.id,
+            'date': today
+        })
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:683]: Same comment, you can do the create in batch. 🙂

**Review by xavierbol — COMMENTED** [review]
Thanks for the fix, I added some comments. 🙂

File: addons/sale_timesheet/models/sale_order.py (hunk 1 above)
Comment (pivi-odoo) [reply addons/sale_timesheet/models/sale_order.py:288]: At first in the mapping the value associated with the line is the delivered_quantity, by removing the quantity that was already invoiced from the delivered_quantity, we get the difference, which is the quantity we want to invoice.
If a line is already fully invoiced, the difference between the delivered_quantity and invoiced_quantity is 0, and when at the line `line.qty_to_invoice = ...`(L:291) is set to 0, the line is not included in the generated invoice.

File: addons/sale_timesheet/models/sale_order.py [hunk 5]
```diff
@@ -280,5 +280,12 @@ def _recompute_qty_to_invoice(self, start_date, end_date):
             domain = expression.AND([domain, [('date', '<=', end_date)]])
         mapping = lines_by_timesheet.sudo()._get_delivered_quantity_by_analytic(domain)
 
+        # we need to remove the already invoiced quantity for lines that were credited
+        if refund_account_moves:
+            for sol_id in mapping:
+                line = refund_account_moves.line_ids.sale_line_ids.filtered(lambda sol: sol.id == sol_id)
```
[the remark this answers is not in the document]
Comment (pivi-odoo) [reply addons/sale_timesheet/models/sale_order.py]: I've changed the filter with a domain search, I hope it addresses your remark :)

File: addons/sale_timesheet/models/sale_order.py [hunk 6]
```diff
@@ -280,5 +280,14 @@ def _recompute_qty_to_invoice(self, start_date, end_date):
             domain = expression.AND([domain, [('date', '<=', end_date)]])
         mapping = lines_by_timesheet.sudo()._get_delivered_quantity_by_analytic(domain)
 
+        # we need to remove the already invoiced quantity for lines that were credited
+        if refund_account_moves:
+            lines = self.env["sale.order.line"].search([
+                ("id", "in", list(mapping)),
+                ("invoice_lines", "in", refund_account_moves.line_ids.ids)
```
Comment (xavierbol) [inline addons/sale_timesheet/models/sale_order.py:287]: linting. 😇 
```suggestion
                ("invoice_lines", "in", refund_account_moves.line_ids.ids),
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 7]
```diff
@@ -524,6 +524,178 @@ def test_timesheet_invoice(self):
...
+                'order_id': sale_order.id,
+            },
+            {
+                'name': self.product_delivery_timesheet3.name,
+                'product_id': self.product_delivery_timesheet3.id,
+                'product_uom_qty': 50,
+                'product_uom': self.product_delivery_timesheet3.uom_id.id,
+                'price_unit': self.product_delivery_timesheet3.list_price,
+                'order_id': sale_order.id,
+            }
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:563]: 😇 
```suggestion
            },
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 8]
```diff
@@ -524,6 +524,178 @@ def test_timesheet_invoice(self):
...
+        invoice_refund.action_post()
+        # Log 5 timesheet hours
+        self.env['account.analytic.line'].create([
+            {
+                'name': 'Test Line 2',
+                'project_id': task_serv1.project_id.id,
+                'task_id': task_serv1.id,
+                'unit_amount': 5,
+                'employee_id': self.employee_manager.id,
+                'date': today
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:627]:
```suggestion
                'date': today,
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 9]
```diff
@@ -524,6 +524,178 @@ def test_timesheet_invoice(self):
...
+            },
+            {
+                'name': 'Test Line 2',
+                'project_id': task_serv2.project_id.id,
+                'task_id': task_serv2.id,
+                'unit_amount': 5,
+                'employee_id': self.employee_manager.id,
+                'date': today
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:635]:
```suggestion
                'date': today,
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 10]
```diff
@@ -524,6 +524,178 @@ def test_timesheet_invoice(self):
...
+        ])
+        # now we should have 35 hours delivered, and 20 already invoiced
+        for line in sale_order.order_line:
+            self.assertEqual(line.qty_delivered, 35, "We should have 35 (30 + 5) hours delivered")
+            self.assertEqual(line.qty_invoiced, 20, "We should have 20 (30 - 10) hours invoiced")
+        # create an invoice that contains the timesheeted days
+        invoice_timesheeted_wizard = self.env['sale.advance.payment.inv'].with_context(self.context).create({
+            'advance_payment_method': 'delivered',
+            'date_start_invoice_timesheet': today,
+            'date_end_invoice_timesheet': today + timedelta(days=1)
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:646]:
```suggestion
            'date_end_invoice_timesheet': today + timedelta(days=1),
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 11]
```diff
@@ -524,6 +524,178 @@ def test_timesheet_invoice(self):
...
+            self.assertEqual(line.qty_delivered, 35, "We should have 35 (30 + 5) hours delivered")
+            self.assertEqual(line.qty_invoiced, 32, "We should have 32 (35 - 3) hours invoiced")
+        # Log the rest of the hours necessary
+        self.env['account.analytic.line'].create([{
+                'name': 'Test Line 2',
+                'project_id': task_serv1.project_id.id,
+                'task_id': task_serv1.id,
+                'unit_amount': 15,
+                'employee_id': self.employee_manager.id,
+                'date': today
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:680]:
```suggestion
                'date': today,
```

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 12]
```diff
@@ -524,6 +524,178 @@ def test_timesheet_invoice(self):
...
+            },
+            {
+                'name': 'Test Line 2',
+                'project_id': task_serv2.project_id.id,
+                'task_id': task_serv2.id,
+                'unit_amount': 15,
+                'employee_id': self.employee_manager.id,
+                'date': today
```
Comment (xavierbol) [inline addons/sale_timesheet/tests/test_sale_timesheet.py:688]:
```suggestion
                'date': today,
```

**Review by xavierbol — COMMENTED** [review]
Thanks again for your work, I have made some comments (mostly for linting) after I will look good to me.

I let @IT-Ideas doing the final review. 🙂

**Review by IT-Ideas — COMMENTED** [review]
Thanks for the fix 🙂 
I am just wondering whether this would be a standard way of doing things in odoo.
Why not adding manually a line in the SO with a quantity set to -2 rather that making this computation behind the scenes? I think this could be confusing for the user to see quantity of 7 in the SOL, 5 as invoiced and when clicking on invoice, nothing is invoiced.

I will ask @aju-odoo 's opinion on this (she is the PO of the Timesheet app).

**Comment by pivi-odoo** [comment]
This issue was classified as a "technical limitation".
We currently don't support partial refund for lines that are a timesheet service, only full refund.
An alternative flow is to do a full refund and then change the quantity on the SOL.
Closing this PR.

**Pull request closed without merging**